arrow-fx-reactor/src/main/kotlin/arrow/fx/reactor/MonoK.kt

```diff
@@ -97,33 +102,64 @@ data class MonoK<out A>(val mono: Mono<out A>) : MonoKOf<A> {
-      val a: A? = mono.block()
-      if (a != null) {
-        if (isCancelled.value) release(a, ExitCase.Cancelled).fix().mono.subscribe({}, sink::error)
-        else try {
-          sink.onDispose(use(a).fix()
-            .flatMap { b -> release(a, ExitCase.Completed).fix().map { b } }
-            .handleErrorWith { e -> release(a, ExitCase.Error(e)).fix().flatMap { MonoK.raiseError<B>(e) } }
-            .mono
-            .doOnCancel { release(a, ExitCase.Cancelled).fix().mono.subscribe({}, sink::error) }
-            .subscribe(sink::success, sink::error)
+    Mono.create<B> { emitter ->
+      val cancelable = ForwardDisposable()
+      emitter.onCancel(cancelable.cancel())
+
+      value().subscribe(
+        { a: A ->
+          cancelable.complete(
+            Mono.defer { use(a).value() }
+              .doOnCancel { Mono.defer { release(a, ExitCase.Cancelled).value() }.subscribe() }
+              .subscribe({ b ->
+                Mono.defer { release(a, ExitCase.Completed).value() }
+                  .subscribe({ emitter.success(b) }, { e -> emitter.error(e) })
+              }, { e ->
+                Mono.defer { release(a, ExitCase.Error(e)).value() }
+                  .subscribe({ emitter.error(e) }, { e2 -> emitter.error(Platform.composeErrors(e, e2)) })
+              }, {
+                Mono.defer { release(a, ExitCase.Completed).value() }
+                  .subscribe({ emitter.success() }, { e -> emitter.error(e) })
+              })
           )
-        } catch (e: Throwable) {
-          if (NonFatal(e)) {
-            release(a, ExitCase.Error(e)).fix().mono.subscribe({
-              sink.error(e)
-            }, { e2 ->
-              sink.error(Platform.composeErrors(e, e2))
-            })
-          } else {
-            throw e
-          }
-        }
-      } else sink.success(null)
-    })
+        },
+        emitter::error,
+        { }
+      )
+    }.k()
+
+  fun guaranteeCase(f: (ExitCase<Throwable>) -> MonoKOf<Unit>): MonoK<A> =
+    Mono.create<A> { emitter ->
+      val cancelable = ForwardDisposable()
+      emitter.onCancel(cancelable.cancel())
+
+      cancelable.complete(value()
+        .doOnCancel { Mono.defer { f(ExitCase.Cancelled).value() }.subscribe() }
+        .subscribe(
+          { a ->
+            Mono.defer { f(ExitCase.Completed).value() }
+              .subscribe({ emitter.success(a) }, { e -> emitter.error(e) })
+          }, { e ->
+          Mono.defer { f(ExitCase.Error(e)).value() }
+            .subscribe({ emitter.error(e) }, { e2 -> emitter.error(Platform.composeErrors(e, e2)) })
+        }, {
+          Mono.defer { f(ExitCase.Completed).value() }
+            .subscribe({ emitter.success() }, { e -> emitter.error(e) })
+        }))
+    }.k()
+
+  fun fork(coroutineContext: CoroutineContext): MonoK<Fiber<ForMonoK, A>> =
+    coroutineContext.asScheduler().let { scheduler ->
+      Mono.create<Fiber<ForMonoK, A>> { emitter ->
```

Comment:
If the emitter is only used to return a value, it's better to use the blocking constructor. In `Rx` is something like `fromCompletable`.

Reply:
Right!